### PR TITLE
Add preflight

### PR DESCRIPTION
**Pull request by aptxkid**
Add preflight interface to `File` class.

I will add preflight to file update and file upload in later PRs.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @aptxkid has signed the CLA. Thanks for the pull request!

**Comment by kapitoshka1812** [comment]
This looks good.

**Comment by aptxkid** [comment]
I am working on unittest.

File: demo/example.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import os
 from boxsdk.client import Client
 from boxsdk.object.collaboration import CollaborationRole
+from boxsdk.object.file import File
```
Comment (Jeff-Meadows) [inline demo/example.py]: Is this the only change to example.py?

File: boxsdk/object/base_endpoint.py [hunk 2]
```diff
@@ -17,7 +17,8 @@ def __init__(self, session):
         """
         self._session = session
 
-    def get_url(self, endpoint, *args):
+    @classmethod
```
Comment (Jeff-Meadows) [inline boxsdk/object/base_endpoint.py]: I don't think it's a good idea to make this a classmethod when subclasses will not make it a classmethod.

File: boxsdk/object/base_object.py [hunk 3]
```diff
@@ -65,14 +65,15 @@ def get_url(self, *args):
         """
         return super(BaseObject, self).get_url('{0}s'.format(self._item_type), self._object_id, *args)
 
-    def get_type_url(self):
+    @classmethod
```
Comment (Jeff-Meadows) [inline boxsdk/object/base_object.py]: ditto

File: demo/example.py (hunk 1 above)
Comment (aptxkid) [inline demo/example.py]: @Jeff-Meadows I tested in example.py briefly. I will remove this line.

File: boxsdk/object/base_endpoint.py (hunk 2 above)
Comment (aptxkid) [inline boxsdk/object/base_endpoint.py]: hmm.. @Jeff-Meadows I agree that get_url() should not be a class method.

File: boxsdk/object/base_object.py [hunk 4]
```diff
@@ -65,14 +65,14 @@ def get_url(self, *args):
         """
         return super(BaseObject, self).get_url('{0}s'.format(self._item_type), self._object_id, *args)
 
-    def get_type_url(self):
+    def get_type_url(self, *args):
```
Comment (aptxkid) [inline boxsdk/object/base_object.py]: I need to remove this change to get_type_url since it's no longer used.

File: boxsdk/object/file.py [hunk 5]
```diff
@@ -11,6 +12,8 @@ class File(Item):
 
     _item_type = 'file'
 
+
```
Comment (aptxkid) [inline boxsdk/object/file.py]: Remove the empty lines.

File: boxsdk/object/item.py [hunk 6]
```diff
@@ -5,11 +5,56 @@
 import json
 
 from .base_object import BaseObject
+from boxsdk.config import API
 
 
 class Item(BaseObject):
     """Box API endpoint for interacting with files and folders."""
 
+    @classmethod
```
Comment (kunalparmar) [inline boxsdk/object/item.py]: Why is this a class method?

Comment (aptxkid) [inline boxsdk/object/item.py]: Preflight is a little different than other API calls in a sense that you can do preflight when trying to update an existing file on Box or uploading a new file. So it doesn't make sense to make a normal method tie to an object. But it can be a staticmethod, if that's why you're asking.

Comment (kunalparmar) [inline boxsdk/object/item.py]: Is uploading a new file a class/static method?

File: boxsdk/object/item.py [hunk 7]
```diff
@@ -5,11 +5,56 @@
 import json
 
 from .base_object import BaseObject
+from boxsdk.config import API
```
Comment (kunalparmar) [inline boxsdk/object/item.py]: Import style is inconsistent with the line above

File: boxsdk/object/item.py [hunk 8]
```diff
@@ -5,11 +5,56 @@
...
+        url = '{0}/files/{1}'.format(API.BASE_API_URL, endpoint)
+        data = {'size': size}
+        if name:
+            data['name'] = name
+        if parent_id:
+            data['parent'] = {'id': parent_id}
+
+        session.options(
+            url=url,
+            expect_json_response=False,
```
Comment (kunalparmar) [inline boxsdk/object/item.py]: I don't see this called out in the developer docs. What does the options call return?

File: boxsdk/object/item.py (hunk 7 above)
Comment (aptxkid) [inline boxsdk/object/item.py]: IMO, this is actually OK. Since 'item' module is in the 'object' package while apparently config module is not. Ideally, IMO, we would add 'from .config import API' to boxsdk/__init__.py and we can just do 'from boxsdk import API' here.

File: test/unit/object/test_item.py [hunk 9]
```diff
@@ -105,3 +108,49 @@ def test_get(test_item_and_response, mock_box_session, fields, mock_object_id, e
...
     assert info.id == mock_object_id
+
+
+@pytest.mark.parametrize(
+    'size, name, file_id, parent_id, expected_url, expected_data',
+    [
+        (
+            # Testing preflight check of a new file to be uploaded to Box.
+            100,
+            'foo.txt',
```
Comment (kunalparmar) [inline test/unit/object/test_item.py]: Rotunicode?

File: test/unit/object/test_item.py [hunk 10]
```diff
@@ -105,3 +108,49 @@ def test_get(test_item_and_response, mock_box_session, fields, mock_object_id, e
...
+            100,
+            None,
+            '123456',
+            None,
+            '{0}/files/123456/content'.format(API.BASE_API_URL),
+            json.dumps({'size': 100}),
+        ),
+    ],
+)
+def test_preflight_check(
```
Comment (kunalparmar) [inline test/unit/object/test_item.py]: You are missing tests for input combinations that may be None.

File: boxsdk/object/item.py (hunk 6 above)
Comment (aptxkid) [inline boxsdk/object/item.py]: Uploading a file is a method of a File object, so it's tied to a file. One option would be making one preflight method to File for new file uploads, and adding one different preflight method to File for updating existing files on Box. In that way, nothing needs to be a class/static method. But we have two preflight methods. But that might be a good idea actually. Since code duplication can be addressed by putting code needs to be shared in Item. I think I like this idea more!

File: boxsdk/object/item.py (hunk 8 above)
Comment (aptxkid) [inline boxsdk/object/item.py]: All we care about is if this call returns 200 or not. If it's not a 200 return, by default, the BoxSession will raise a BoxAPIException.

File: test/unit/object/test_item.py (hunk 9 above)
Comment (aptxkid) [inline test/unit/object/test_item.py]: We are simply checking if we are making the call with the same name. But sure! :)

File: test/unit/object/test_item.py (hunk 10 above)
Comment (aptxkid) [inline test/unit/object/test_item.py]: I believe these are all the combinations that Box supports. I can double check.

File: boxsdk/object/item.py (hunk 7 above)
Comment (kunalparmar) [inline boxsdk/object/item.py]: You can do a relative import as from ..config import API

File: boxsdk/object/item.py (hunk 8 above)
Comment (kunalparmar) [inline boxsdk/object/item.py]: Is this required then?

File: boxsdk/object/item.py (hunk 6 above)
Comment (kunalparmar) [inline boxsdk/object/item.py]: That sounds like a more consistent design choice to me.

Comment (Jeff-Meadows) [inline boxsdk/object/item.py]: Yes, +1 to this.

File: test/unit/object/test_file.py [hunk 11]
```diff
@@ -99,3 +99,43 @@ def test_unlock(test_file, mock_box_session, mock_file_response):
...
+        kwargs['name'] = name
+    test_file.preflight_check(**kwargs)
+    mock_box_session.options.assert_called_once_with(
+        url='{0}/files/{1}/content'.format(
+            API.BASE_API_URL,
+            mock_object_id,
+        ),
+        expect_json_response=False,
+        data=expected_data,
+    )
```
Comment (Jeff-Meadows) [inline test/unit/object/test_file.py]: add newline

File: test/unit/object/test_folder.py [hunk 12]
```diff
@@ -167,3 +167,19 @@ def test_update_sync_state(test_folder, mock_folder_response, mock_box_session,
     mock_box_session.put.assert_called_once_with(expected_url, data=json.dumps(data), params=None, headers=None)
     assert isinstance(update_response, Folder)
     assert update_response.object_id == test_folder.object_id
+
+
+def test_preflight(test_folder, mock_object_id, mock_box_session):
```
Comment (Jeff-Meadows) [inline test/unit/object/test_folder.py]: Parametrize file size/name.

File: boxsdk/object/item.py [hunk 13]
```diff
@@ -5,11 +5,51 @@
...
+        """
+        Make an API call to check if certain file can be uploaded to Box or not.
+        (https://developers.box.com/docs/#files-preflight-check)
+
+        :param size:
+            The size of the file in bytes. Specify 0 for unknown file-sizes.
+        :type size:
+            `int`
+        :param name:
+            The name of the file to be uploaded. This is optional if `file_id` is specified.
```
Comment (Jeff-Meadows) [inline boxsdk/object/item.py]: end line with comma and start next line with lowercase but

File: boxsdk/object/item.py [hunk 14]
```diff
@@ -5,11 +5,51 @@
...
 class Item(BaseObject):
     """Box API endpoint for interacting with files and folders."""
 
+    def _preflight_check(self, size, name=None, file_id=None, parent_id=None):
+        """
+        Make an API call to check if certain file can be uploaded to Box or not.
+        (https://developers.box.com/docs/#files-preflight-check)
+
+        :param size:
+            The size of the file in bytes. Specify 0 for unknown file-sizes.
```
Comment (Jeff-Meadows) [inline boxsdk/object/item.py]: The file to be uploaded.

Change to "unknown file size"

File: test/unit/object/test_folder.py (hunk 12 above)
Comment (aptxkid) [inline test/unit/object/test_folder.py]: Any suggestions about exactly how to parametrize size/name? I did it for preflight in File because 'name' can be omitted.

File: test/unit/object/test_file.py (hunk 11 above)
Comment (aptxkid) [inline test/unit/object/test_file.py]: Done

**Comment by Jeff-Meadows** [comment]
LGTM after travis is done.
